langs/java.go

```diff
@@ -184,50 +190,121 @@ func (h *JavaLangHelper) getFDKAPIVersion() (string, error) {
+	const bintrayVersionURL = "https://api.bintray.com/search/packages/maven?repo=fnproject&g=com.fnproject.fn&a=fdk"
+	const mavenVersionUrl = "https://repo1.maven.org/maven2/com/fnproject/fn/fdk/maven-metadata.xml"
 
-	type parsedResponse struct {
-		Version string `json:"latest_version"`
-	}
+	const versionEnv = "FN_JAVA_FDK_VERSION"
+	fetchError := fmt.Errorf("Failed to fetch latest Java FDK javaVersion. Check your network settings or manually override the javaVersion by setting %s", versionEnv)
 	version := os.Getenv(versionEnv)
+
 	if version != "" {
 		return version, nil
 	}
+	version, err := h.getFDKLastestFromURL(mavenVersionUrl, bintrayVersionURL)
+	if err != nil {
+		return "", fetchError
+	}
 
-	// nishalad95: bin tray TLS certs cause verification issues on OSX, skip TLS verification
+	h.latestFdkVersion = version
+	return version, nil
+}
+
+func (h *JavaLangHelper) getFDKLastestFromURL(comURL string, bintrayURL string) (string, error) {
+	var buf *bytes.Buffer
+	var err error
+	err = fmt.Errorf("All URL failed to respond ")
+
+	//First search for com.fnproject.fn from Maven Central to get the latest version
+	buf, err = h.getURLResponse(comURL, false)
+	if err == nil {
+		version, e1 := h.parseMavenResponse(*buf)
+		if e1 == nil {
+			h.pomType = "maven"
+			return version, e1
+		}
+	}
+
+	//Second time search for com.fnproject.fn from Bintray to get the latest version, if fetch from Maven fails
+	buf, err = h.getURLResponse(bintrayURL, true)
+	if err == nil {
+		version, e1 := h.parseBintrayResponse(*buf)
+		if e1 == nil {
+			h.pomType = "bintray"
+			return version, e1
+		}
+	}
+
+	//In all other case return error as latest FDK version is not identified
+	return "", err
+}
+
+func (h *JavaLangHelper) getURLResponse(url string, inSecureSkipVerify bool) (*bytes.Buffer, error) {
 	defaultTransport := http.DefaultTransport.(*http.Transport)
+	// nishalad95: bin tray TLS certs cause verification issues on OSX, skip TLS verification
 	noVerifyTransport := &http.Transport{
 		Proxy:                 defaultTransport.Proxy,
 		DialContext:           defaultTransport.DialContext,
 		MaxIdleConns:          defaultTransport.MaxIdleConns,
 		IdleConnTimeout:       defaultTransport.IdleConnTimeout,
 		ExpectContinueTimeout: defaultTransport.ExpectContinueTimeout,
 		TLSHandshakeTimeout:   defaultTransport.TLSHandshakeTimeout,
-		TLSClientConfig:       &tls.Config{InsecureSkipVerify: true},
+		TLSClientConfig:       &tls.Config{InsecureSkipVerify: inSecureSkipVerify},
 	}
 	client := &http.Client{Transport: noVerifyTransport}
-
-	resp, err := client.Get(versionURL)
+	resp, err := client.Get(url)
 	if err != nil || resp.StatusCode != 200 {
-		return "", fetchError
+		return nil, fmt.Errorf("Failed to fetch response from URL %s Error: %v Status: %d", url, err, resp.StatusCode)
 	}
-
-	buf := bytes.Buffer{}
+	buf := &bytes.Buffer{}
```

Comment:
Consider converting this into a byte array instead of a buffer? We have a `io.Reader` (resp.Body) here and we are converting it into bytes.Buffer and then to Bytes in child functions.

Reply:
DOne